src/dbnode/encoding/proto/README.md

```diff
@@ -0,0 +1,269 @@
+
+1. Write the full string "foo" into the stream and add "foo" to the LRU -> `["foo"]`
+2. Write the full string "bar" into the stream and add "bar" to the LRU -> `["foo", "bar"]`
+3. Write the full string "baz" into the stream and and evict "foo" from the LRU -> `["bar", "baz"]`
+4. Encode index 0 into the stream with a single bit (representing the string "bar") and update the LRU -> `["baz", "bar"]`
+
+This compression scheme works because the decoder can maintain its own LRU (of the same maximum capacity) and apply the same operations in the same order when its decompressing the stream.
+As a result, when it encounters an encoded LRU index it can lookup the corresponding string in its own LRU at the specified index.
+
+##### Encoding
+
+The LRU dictionary compression scheme uses 3 control bits to encode all the relevant information required to decode the stream.
+
+The first control bit is the "no change" control bit.
+If this is set to `1` then the value has not changed since its previous value and no further encoding/decoding is require.
+
+The second control bit determines how we will interpret the subsequent bits.
+If its `0` then we interpret the next `n` bits (where `n` is the number of bits required to encode the highest index of the configured LRU capacity) as the index into the LRU of recently encoded / decoded strings which this value corresponds to.
+
+If the second control bit is set to `1` then the remainder should be interpreted as a `varint` encoding the length of the `bytes` followed by the `bytes` themselves.
+One caveat to keep in mind is that if after encoding the length `varint` the encode is not aligned on a byte boundary we'll pad the stream with `0`s until the next byte.
+This isn't a strict requirement of the encoding scheme (in fact it reduces compression slightly), but it greatly simplifies the implementation because the encoder needs to be able to reference bytes that its already encoded so it can check if the bytes that are about to be encoded are in the LRU.
+The encoder could keep track of all the bytes that correspond to each LRU entry in memory, but that would be wasteful in terms of memory utilization.
+Instead, its more efficient if the LRU stores offsets into the encoded stream for the beginning and end of the bytes that have already been encoded.
+These offsets are much easier to keep track of and compare against if we can assume they always correspond to the beginning of a byte boundary and we don't have to keep track of the bit offset into an individual byte as well.
+In the future we may change the implement to favor complicating the code so that we don't have to waste any bits on the padding.
+
+### Compression Limitations
+
+This package will attempt to compress all scalar type fields at the top level of a message, but will not compress any data that is part of a `repeated`, `map` or `nested message` field.
```

Comment:
Compressing a repeated field well would just be really complicated because you'd have to account for changes in the length of the repeated sequence from one message to the next. We could do it, but I wouldn't want to support it unless we had a really good use case that was difficult to model any other way.

Maps are even more difficult because you'd have to deal with the sorting issue, as well as the fact that keys may be added and removed arbitrarily. It would get messy really quickly.